### PR TITLE
[pipelining] Optional per-direction P2P communicators for PipelineStage

### PR DESCRIPTION
**Pull request by tushar00jain**
Summary:
Pipeline parallelism issues all send/recv on a single communicator.

NCCL-group coalescing (batch_isend_irecv -> one ncclGroupStart/End) makes a single mixed
send+recv batch deadlock-free, but it does not remove ordering hazards across
separate batches.

This adds per-direction PP communicators. When enabled, forward (
r->r+1) and backward (r->r-1) P2P are carried on two separate
communicators derived from the stage's group via split_group.

Test Plan:
- `test_pp_per_direction_pipeline_stage.py`
- `test_pp_torchcomms_p2p_deadlock.py`

### CODE REVIEW

**Comment by d4l3k** [comment]
@claude review this PR

File: torch/distributed/pipelining/schedules.py [hunk 1]
```diff
@@ -633,6 +633,18 @@ def _batch_p2p(p2p_ops: list[dist.P2POp], desc: str | None = None) -> list[dist.
...
+    # Per-direction P2P (config.pipeline_per_direction_p2p) tags forward and
+    # backward ops with different communicators. A fused batch (e.g. 1F1B's
+    # fwd_sends + bwd_recvs) then spans >1 group; issue each group's ops as their
+    # own batch so they run on separate comms/streams instead of one FIFO. When
+    # all ops share a group (the default), this is a no-op fast path.
+    op_groups = {id(p.group): p.group for p in p2p_ops}
+    if len(op_groups) > 1:
+        works: list[dist.Work] = []
+        for g in op_groups.values():
+            works += _batch_p2p([p for p in p2p_ops if p.group is g], desc=desc)
```
Comment (d4l3k) [inline torch/distributed/pipelining/schedules.py]: nit: can we use itertools.groupby so this isn't an O(n^2) operation? in practice it probably doesn't matter but makes it easier to understand what's going on here

File: torch/distributed/pipelining/stage.py [hunk 2]
```diff
@@ -126,6 +126,64 @@ def __repr__(self):
...
+        )
+        bwd = dist.split_group(
+            parent_pg=group, split_ranks=split_ranks, group_desc="pp_p2p_bwd"
+        )
+    except RuntimeError as err:
+        raise RuntimeError(
+            "Per-direction pipeline P2P needs the default process group to be "
+            "device-bound so split_group can duplicate the PP communicator. Pass "
+            f"device_id to init_process_group. Original error: {err}"
+        ) from err
```
Comment (d4l3k) [inline torch/distributed/pipelining/stage.py]: Lets not use RuntimeError for this. If there's a device-bound error can we change the error type/binding to produce a more closely typed error (i.e. DeviceError) ? Same way we have c10d specific socket/timeout errors https://github.com/pytorch/pytorch/blob/423814f4e6aea217908c9e0323da224735f360f9/torch/csrc/distributed/c10d/exception.h#L36-L38

Or can we check a device property on the group instead?

File: torch/distributed/config.py [hunk 3]
```diff
@@ -27,6 +27,19 @@
...
 
+# When enabled, pipeline stages carry forward (downstream) and backward
+# (upstream) P2P on two separate communicators instead of sharing one. A single
+# PP communicator serializes all send/recv in one FIFO: coalescing makes a single
+# mixed batch deadlock-free, but across batches (pipeline skew, looped / V
+# schedules, skip connections) the shared FIFO can still form a dependency cycle
+# and deadlock. Splitting by direction removes that hazard and restores
+# full-duplex bandwidth. PipelineStage(p2p_per_direction=...) overrides this
+# default per stage. Requires a device-bound default process group.
+pipeline_per_direction_p2p: bool = Config(
```
Comment (d4l3k) [inline torch/distributed/config.py:42]: should we default this to true if torchcomms is being used?

File: test/distributed/pipelining/test_pp_torchcomms.py [hunk 4]
```diff
@@ -0,0 +1,351 @@
+# Owner(s): ["oncall: distributed"]
```
Comment (d4l3k) [inline test/distributed/pipelining/test_pp_torchcomms.py]: naming the file after the expected error is a bit odd -- maybe just name this test_pp_torchcomms.py

**Review by d4l3k — APPROVED** [review]
Code seems reasonable but would be nice to have someone with more familiarity with pipelining take a look. I only skimmed the tests, the core logic seems correct

File: torch/distributed/pipelining/stage.py [hunk 5]
```diff
@@ -126,6 +126,64 @@ def __repr__(self):
...
+# that share one PP group; they must share the same forward/backward comms (and
+# issue the split_group collective only once) so creation stays consistent and
+# cheap across all ranks.
+_PP_DIRECTION_GROUP_CACHE: dict[Any, tuple[dist.ProcessGroup, dist.ProcessGroup]] = {}
+
+
+def _build_p2p_direction_groups(
+    group: dist.ProcessGroup | None,
+) -> tuple[dist.ProcessGroup, dist.ProcessGroup]:
+    """Create two communicators over the same ranks as ``group`` for the forward
```
Comment (wconstab) [inline torch/distributed/pipelining/stage.py]: nit: the naming choice (fwd, bwd) is too aliased with the fwd/bwd pass and its hard to reinterpret it as 'right neighbor/ left neighbor' - suggest alternate naming

File: torch/distributed/pipelining/stage.py [hunk 6]
```diff
@@ -167,6 +225,21 @@ def __init__(
         self.device = device
         self.group = group
 
+        # Forward (downstream) and backward (upstream) P2P communicators. Enabled
+        # globally via torch.distributed.config.pipeline_per_direction_p2p (env
+        # TORCH_DISTRIBUTED_PIPELINE_PER_DIRECTION_P2P); when disabled both alias
+        # ``self.group`` so behavior is byte-for-byte unchanged.
+        import torch.distributed.config as dist_config
```
Comment (wconstab) [inline torch/distributed/pipelining/stage.py]: is this a circular import? or can you move it to the top?

File: torch/distributed/distributed_c10d.py [hunk 7]
```diff
@@ -5834,16 +5798,31 @@ def new_group(
...
+        # FakeProcessGroup directly (see ``_new_group_with_tag``). When the
+        # requested backend matches the parent (including a fake parent), split
+        # delegation is fine.
+        parent_backend, _ = _world.pg_map[_get_default_group()]
+        is_fake_subgroup = (
+            backend is not None
+            and str(backend).lower() == "fake"
+            and str(backend).lower() != str(parent_backend).lower()
         )
+        if not is_fake_subgroup:
```
Comment (wconstab) [inline torch/distributed/distributed_c10d.py:5863]: unclear to me how fake_subgroup is tied to the other changes on this PR

**Review by wconstab — COMMENTED** [review]
meta point- hard to know how much validation this PR needs. have we done any >8gpu test runs with the torchcomms path?

File: torch/distributed/distributed_c10d.py (hunk 7 above)
Comment (tushar00jain) [reply torch/distributed/distributed_c10d.py:5863]: this is a different commit - the previous pr on the stack

**Comment by tushar00jain** [comment]
> meta point- hard to know how much validation this PR needs. have we done any >8gpu test runs with the torchcomms path?

did a 4 gpu test with titan

File: torch/distributed/pipelining/schedules.py [hunk 8]
```diff
@@ -633,6 +633,20 @@ def _batch_p2p(p2p_ops: list[dist.P2POp], desc: str | None = None) -> list[dist.
     desc_str = f"{desc}, " if desc else ""
     logger.debug("batch_p2p %s%s", desc_str, p2p_ops)
 
+    # Per-direction P2P (config.pipeline_per_direction_p2p) tags forward and
+    # backward ops with different communicators. A fused batch (e.g. 1F1B's
+    # fwd_sends + bwd_recvs) then spans >1 group; issue each group's ops as their
+    # own batch so they run on separate comms/streams instead of one FIFO. When
+    # all ops share a group (the default), this is a no-op fast path.
+    ops_by_group: dict[int, list[dist.P2POp]] = {}
```
Comment (sanketpurandare) [inline torch/distributed/pipelining/schedules.py]: `_batch_p2p` now chooses the issue order for the forward/backward process groups from the first occurrence order in `p2p_ops`. In the real 1F1B path, the two sides of the same neighbor exchange can naturally call this with opposite list order (`fwd_sends + bwd_recvs` on one side and `bwd_sends + fwd_recvs` on the other), so adjacent ranks can issue the two NCCL communicators in different orders. That is exactly the kind of cross-communicator ordering hazard this feature needs to prevent or explicitly validate. Please make the split use a deterministic direction ordering, or otherwise document/prove why this ordering is safe, and add a regression that exercises the actual `PipelineStage` + `_batch_p2p` schedule path.

File: test/distributed/pipelining/test_pp_torchcomms.py [hunk 9]
```diff
@@ -0,0 +1,351 @@
...
+    """The fix (mechanism-level): split the same transfers across two
+    per-direction communicators -- ``down`` carries r->r+1 traffic, ``up`` carries
+    r->r-1 traffic -- so forward and backward P2P live on separate comms/streams
+    and cannot block each other in a FIFO.
+
+    This mirrors what PipelineStage does upstream when
+    config.pipeline_per_direction_p2p is set; here we build the two comms with
+    plain ``dist`` primitives (new_group over all ranks, equivalent to split_group
+    for this pure-PP ring) so the test stays self-contained.
+    """
```
Comment (sanketpurandare) [inline test/distributed/pipelining/test_pp_torchcomms.py]: The per-direction ring test creates two plain `dist.new_group()` communicators and manually issues `down` before `up` on every rank. That bypasses the implementation this PR is adding: `PipelineStage`'s `split_group` cache, direction assignment, and `_batch_p2p`'s per-group split/order. As written, the test can pass even if the actual pipeline path still has an ordering bug or fails to wire the derived groups correctly. Please add or replace this with a regression that enables `torch.distributed.config.pipeline_per_direction_p2p` and runs the schedule through `PipelineStage`/`_batch_p2p` in the configuration that used to hang.

File: test/distributed/pipelining/test_pp_torchcomms.py [hunk 10]
```diff
@@ -0,0 +1,351 @@
...
+    def test_torchcomms_native_batch_completes(self):
+        # The primitive the coalescing path routes to: BatchSendRecv is
+        # deadlock-free for the same mixed exchange.
+        result = _spawn_and_watch(
+            "torchcomms", world_size=2, timeout_s=45, native_batch=True
+        )
+        self.assertEqual(result, "completed")
+
+    @requires_gpus(3)
+    def test_ring_single_comm_deadlocks_nccl(self):
```
Comment (sanketpurandare) [inline test/distributed/pipelining/test_pp_torchcomms.py]: `test_ring_single_comm_deadlocks_*` treats a timeout as the successful outcome, so a passing run spends 40s per backend waiting for a hang. It also classifies "some workers crashed and the rest are stuck in collectives/barrier" as `"deadlock"` because `_watch()` checks for alive processes before checking nonzero exit codes. Please avoid landing a regular CI test whose success path is a hang; keep the negative reproducer as a local/debug helper or make it a controlled mechanism test, and use the watchdog around the positive per-direction regression instead.

File: test/distributed/pipelining/test_pp_per_direction_pipeline_stage.py [hunk 11]
```diff
@@ -0,0 +1,322 @@
...
+    HAS_TORCHCOMMS = True
+except ImportError:
+    HAS_TORCHCOMMS = False
+
+
+DIM = 16
+MICRO_BATCH = 4
+N_MICROBATCHES = 4
+GLOBAL_BATCH = MICRO_BATCH * N_MICROBATCHES
+MASTER_PORT = 29613
```
Comment (sanketpurandare) [inline test/distributed/pipelining/test_pp_per_direction_pipeline_stage.py]: This file hard-codes `MASTER_PORT = 29613` and then hand-spawns processes. That is fragile under parallel test execution and duplicates the distributed test harness already used by pipelining tests. Please use the existing `MultiProcContinuousTest`/common distributed helpers in `test_schedule_multiproc.py`, or allocate a free port in the parent and pass it to workers. The same applies to the second new test file's hard-coded default port at `test/distributed/pipelining/test_pp_torchcomms.py:59`.

File: test/distributed/pipelining/test_pp_per_direction_pipeline_stage.py [hunk 12]
```diff
@@ -0,0 +1,322 @@
...
+
+
+def requires_gpus(n: int):
```
Comment (sanketpurandare) [inline test/distributed/pipelining/test_pp_per_direction_pipeline_stage.py]: The custom `requires_gpus()` decorator only checks CUDA availability/device count, but every test in this file initializes `backend="nccl"`. Existing distributed tests use `requires_accelerator_dist_backend(["nccl", "xccl"])` or `requires_nccl()` plus `skip_if_lt_x_gpu()` so they skip correctly when CUDA exists but the distributed accelerator backend is unavailable. Please reuse that infrastructure instead of open-coding the skip logic. The same issue exists in `test/distributed/pipelining/test_pp_torchcomms.py:51`.

**Review by sanketpurandare — CHANGES_REQUESTED** [review]

**Comment by sanketpurandare** [comment]
@claude Conduct a thorough review of this PR. 1) Check for correctness and bugs 2) check for validation/prevention agains't deadlocks when using separate process groups 3) check if main logic is minimal and clean  4) For tests, see if they are efficient (for ex. using MultiProcContinuousTest when possible) and maximally use existing test infrastructure. Tests should comprehensively test functionality yet should be minimal and compact 5) Suggest required changes

File: test/distributed/pipelining/test_pp_torchcomms.py (hunk 10 above)
Comment (tushar00jain) [reply test/distributed/pipelining/test_pp_torchcomms.py]: removed the whole thing, might turn it into a script later I suppose outside of pytorch

**Review by sanketpurandare — CHANGES_REQUESTED** [review]
I don't think we should add `test/distributed/pipelining/test_pp_per_direction_pipeline_stage.py` as a new standalone test file. Most of the file is recreating infrastructure that already exists in the pipelining test suite: process spawning, model/loss setup, schedule construction, result collection, timeouts, and gradient/loss validation. The existing `test_schedule_multiproc.py` checks, which already compare outputs/losses/parameter gradients against a reference model across real `PipelineStage` schedule runs.

  Could you please fold this coverage into the existing tests instead?

 Concretely, I think:
  - the `_batch_p2p` multi-group behavior belongs in `test_schedule.py` next to the existing `TestBatchP2P` tests;
  - the real schedule coverage should be added to `test_schedule_multiproc.py`, likely by running an existing `Schedule1F1B`/manual-stage gradient test with `torch.distributed.config.pipeline_per_direction_p2p=True`;
  - the activation check that `PipelineStage` creates distinct downstream/upstream groups can be a small focused test in the existing multiprocess schedule/stage test infra.

 I understand the new path needs `init_process_group(..., device_id=...)` so that `split_group` can be used. I would prefer adding a small opt-in hook or subclass override in the existing `MultiProcContinuousTest`-based pipelining tests for that, rather than introducing a separate raw `multiprocessing` harness. I would not change `MultiProcContinuousTest._init_pg` unconditionally because that base class also has CPU/Gloo users, and `device_id` changes PG initialization semantics even for accelerator tests.

File: test/distributed/test_fake_backend_new_group.py
Comment (sanketpurandare) [inline test/distributed/test_fake_backend_new_group.py:30]: This test should not set `MASTER_PORT` at all. It is single-process and does not need `env://`; pass s`tore=dist.HashStore()` or a temporary FileStore to `init_process_group(`) instead. That removes the fixed-port collision risk and matches the nearby fake-PG test patterns.

**Review by sanketpurandare — APPROVED** [review]
Please update the PR description

File: test/distributed/test_fake_backend_new_group.py
Comment (tushar00jain) [reply test/distributed/test_fake_backend_new_group.py:30]: this test was in a previous pr on the stack which has already been merged, i can create another pr to fix it
